Updated API to v2 in user mode (issue #92)

pcantera opened the pull request:
Closes #92

<!--- Provide a general summary of your changes in the Title above -->

# Description

<!--- Describe your changes in detail -->

# Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

# Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

# How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

# Screenshots (if appropriate)

# Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

pcantera committed:
Updated API to v2 in user mode (issue #92)

Closes #92

pbanos commented on api/blueprint/services_api.go:
```diff
@@ -29,7 +29,7 @@ func NewServicesService(concertoService utils.ConcertoService) (*ServicesService
 	log.Debug("GetServiceList")
 
-	data, status, err := ss.concertoService.Get("/v1/blueprint/services")
+	data, status, err := ss.concertoService.Get("/v2/blueprint/services")
```
This route is one of the ones that has been discontinued. What you should do is turn the service/service_versions commands into cookbook_versions commands that work with /v2/blueprint/cookbook_versions routes

pcantera replied:
Done

pbanos commented on api/blueprint/templates_api.go:
```diff
@@ -89,7 +89,7 @@ func (tp *TemplateService) CreateTemplate(templateVector *map[string]interface{}
 func (tp *TemplateService) UpdateTemplate(templateVector *map[string]interface{}, ID string) (template *types.Template, err error) {
```
We are going to need a new template compile command that makes use of the new 'PUT /v2/templates/:id/compile' resource

pcantera replied:
Done

pbanos commented on api/dns/domain_api.go:
```diff
@@ -29,7 +29,7 @@ func NewDomainService(concertoService utils.ConcertoService) (*DomainService, er
 	log.Debug("GetDomainList")
 
-	data, status, err := dm.concertoService.Get("/v1/dns/domains")
+	data, status, err := dm.concertoService.Get("/v2/dns/domains")
```
Actually, these resources no longer exist. I believe we already removed the related commands, but it would be a good time to remove the dns-related services, since it makes no sense to update the version on their routes...

pbanos commented on api/network/load_balancers_api.go:
```diff
@@ -28,7 +28,7 @@ func NewLoadBalancerService(concertoService utils.ConcertoService) (*LoadBalance
 	log.Debug("GetLoadBalancerList")
 
-	data, status, err := lb.concertoService.Get("/v1/network/load_balancers")
+	data, status, err := lb.concertoService.Get("/v2/network/load_balancers")
```
The same applies to load balancers, they are not an available resource either...

pbanos commented on api/types/servers.go:
```diff
@@ -25,9 +27,23 @@ type Dns struct {
+	ExecutionOrder  int              `json:"execution_order" header:"EXECUTION_ORDER" show:"noshow,nolist"`
+}
+
+type ScriptCharResponse struct {
```
The response for an execution is actually an event, I think you should reuse the event type for it...

pcantera replied:
Done

pbanos reviewed (changes requested):
Besides what's commented below, regarding the template resource, we probably need to rename the service_list attribute/ service-list flag as run_list/run-list, and add a way to specify the new cookbook_versions

pcantera committed:
Use event type instead new definition (issue #92)

pcantera committed:
Remove obsolete resources (issue #92)

Removed:
- load_balancer
- saas_account
- dns
- domain

pcantera committed:
Remove obsolete setup option (issue #92)

pcantera committed:
Fix time text formatter (issue #92)

pcantera committed:
Added template compile option (issue #92)

pcantera committed:
Added cookbook versions functionalities (issue #92)

Closes #92

pcantera committed:
Fix several log format (issue #92)

Closes #92

pcantera committed:
Apply canonical URLs (issue #92)

Closes #92

pbanos commented on api/types/bootstrapping.go:
```diff
@@ -5,24 +5,24 @@ import (
+	StartedAt             string `json:"started_at,omitempty" header:"STARTED_AT"`
+	FinishedAt            string `json:"finished_at,omitempty" header:"FINISHED_AT"`
+	PolicyfileRevisionIDs string `json:"policyfile_revision_ids,omitempty" header:"POLICY_FILE_REVISION_IDS" show:"nolist"`
+	AttributeRevisionID   string `json:"attribute_revision_id,omitempty" header:"ATTRIBUTE_REVISION_ID"`
```
Do we really need all the header struct tags? I think they are only used for in the user-scope while these types belong exclusively to the server scope...

pcantera replied:
Yes you are right. They are not needed but I think this not harmful; in some way it enriches the develop phase or debugging on a server-if case-

User mode or Server mode, in some way act as client to IMCO. So i prefer to maintain these headers for all objects as possible

pbanos commented on api/types/cookbook_versions.go:
```diff
@@ -0,0 +1,17 @@
+	Destroyable       bool     `json:"destroyable" header:"DESTROYABLE" show:"noshow,nolist"`
+	GlobalLegacy      bool     `json:"global_legacy" header:"GLOBAL_LEGACY" show:"noshow,nolist"`
+	UploadURL         string   `json:"upload_url" header:"UPLOAD_URL" show:"noshow,nolist"`
+	ErrorMessage      string   `json:"error_message" header:"ERROR_MESSAGE" show:"noshow,nolist"`
```
The ErrorMessage, PubliclyAvailable and GlobalLegacy fields should be shown if available, I expect their `show:"noshow,nolist"` tags will prevent it...

pcantera replied:
ok, they will be shown but not be listed
*Edited: Done

pbanos commented on api/types/cookbook_versions.go:
```diff
@@ -0,0 +1,17 @@
+	Recipes           []string `json:"recipes"  header:"RECIPES" show:"nolist"`
+	ResourceType      string   `json:"resource_type" header:"RESOURCE_TYPE" show:"nolist"`
+	PubliclyAvailable bool     `json:"publicly_available" header:"PUBLICLY_AVAILABLE" show:"noshow,nolist"`
+	Destroyable       bool     `json:"destroyable" header:"DESTROYABLE" show:"noshow,nolist"`
```
Remove this field, this was included by mistake and is not part of the public API

pcantera replied:
Done

pbanos commented on api/types/cookbook_versions.go:
```diff
@@ -0,0 +1,17 @@
+	GlobalLegacy      bool     `json:"global_legacy" header:"GLOBAL_LEGACY" show:"noshow,nolist"`
+	UploadURL         string   `json:"upload_url" header:"UPLOAD_URL" show:"noshow,nolist"`
+	ErrorMessage      string   `json:"error_message" header:"ERROR_MESSAGE" show:"noshow,nolist"`
+}
```
Also, cookbook versions are labelable, though I'm not sure that feature is available in the UI...

pcantera replied:
uhm, if possible from ui and IMCO; It could be implemented. I will re-check

pcantera replied:
Done.

pbanos commented on api/types/templates.go:
```diff
@@ -8,21 +8,24 @@ import (
+	ConfigurationAttributes *json.RawMessage `json:"configuration_attributes,omitempty" header:"CONFIGURATION_ATTRIBUTES" show:"nolist"`
 	ResourceType            string           `json:"resource_type" header:"RESOURCE_TYPE" show:"nolist"`
+	CookbookVersions        *json.RawMessage `json:"cookbook_versions,omitempty" header:"COOKBOOK_VERSIONS" show:"nolist"`
+	Variables               *json.RawMessage `json:"variables,omitempty" header:"VARIABLES" show:"nolist"`
```
This is not an input parameter, but an output-only IMCO-calculated one... could you model it as a `map[string]interface{}`?

pcantera replied:
Not sure how to retrieve a valid value for debugging. So far, It's always empty

pbanos replied:
If you deploy a CAT that using dynamic attributes (the 2-tier magento CAT for instance) you should get a template with some variables. Anyway, it is not important to show this and maybe we should remove the field from this CLI model for the time being...

pcantera replied:
Done, removed

pbanos commented on api/types/templates.go:
```diff
@@ -8,21 +8,24 @@ import (
+	RunList                 []string         `json:"run_list,omitempty" header:"RUN_LIST" show:"nolist"`
+	ConfigurationAttributes *json.RawMessage `json:"configuration_attributes,omitempty" header:"CONFIGURATION_ATTRIBUTES" show:"nolist"`
 	ResourceType            string           `json:"resource_type" header:"RESOURCE_TYPE" show:"nolist"`
+	CookbookVersions        *json.RawMessage `json:"cookbook_versions,omitempty" header:"COOKBOOK_VERSIONS" show:"nolist"`
```
We should think of some scheme so that the input of cookbook versions is user-friendlier... Similar to what was proposed for firewall profile rules.

pcantera replied:
Yes, the idea is to improve that in close future along next releases. Focusing in these kind of fields.

pbanos commented on api/types/templates.go:
```diff
@@ -8,21 +8,24 @@ import (
-	ConfigurationAttributes *json.RawMessage `json:"configuration_attributes,omitempty" header:"CONFIGURATION ATTRIBUTES" show:"nolist"`
+	GenericImageID          string           `json:"generic_image_id,omitempty" header:"GENERIC_IMAGE_ID"`
+	RunList                 []string         `json:"run_list,omitempty" header:"RUN_LIST" show:"nolist"`
+	ConfigurationAttributes *json.RawMessage `json:"configuration_attributes,omitempty" header:"CONFIGURATION_ATTRIBUTES" show:"nolist"`
```
Regarding configuration attributes, json.RawMessage is ok, but it would be nice if we could take a file's content (or even STDIN) as data

pcantera replied:
uhm, this could be done in next release while evaluating these kind of fields (as cookbook version, firewall rule, etc)

pbanos commented on cmd/template_cmd.go:
```diff
@@ -91,15 +91,16 @@ func TemplateCreate(c *cli.Context) error {
 	checkRequiredFlags(c, []string{"name", "generic_image_id"}, formatter)
 	// parse json parameter values
-	params, err := utils.FlagConvertParamsJSON(c, []string{"service_list", "configuration_attributes"})
+	params, err := utils.FlagConvertParamsJSON(c, []string{"run_list", "cookbook_versions", "configuration_attributes"})
```
In understand this means the run_list must be specified as a JSON array... If that is the case, a comma-separated format would be nicer: `--run-list mycookbook::a,othercookbook,wordpress::default`

pcantera replied:
Sure, by now, aligned with previous similar functionality.
But it would nice to improve all of these composite types

pcantera replied:
Done, looking for a more friendly mode for user.

In addition, @pbanos , Are you interested in renaming --run_list command to --run-list?

pbanos replied:
Yes, but not only for the run list: in general, dashes as word separators should be preferred for CLI flags over underscores. This is something that we do not want to address necessarily in this PR, though.

pbanos commented on utils/exec.go:
```diff
@@ -45,19 +45,19 @@ func ExecCode(code string, path string, filename string) (output string, exitCod
 	os.Chmod(tmp.Name(), 0777)
 	if err != nil {
-		log.Fatalf("Error changing permision to file : ", err)
+		log.Fatalf("Error changing permission to file: %v", err)
```
Where is ExecCode called? `log.Fatalf` calls will force an exit of the process, so if this is used in command-polling or bootstrapping commands, we should remove them...

pcantera replied:
Just to fix some bad use detected while testing with coverage.
It's used in Server Mode scripts: "Manages Execution Scripts within a Host"... (*code to be re-factored in another stage)
*for "executeScriptCharacterization"(boot|operational|shutdown)

pbanos reviewed (commented):
Here are some comments, some ideas for future work and other concerns about the implementation but if you can dismiss them it is good to go! Good work! :+1:

pcantera committed:
Fix cookbook version properties (issue #92)

Closes #92

pcantera committed:
Fix labelable (issue #92)

 - Added labelable functionality for cookbook version
 - Fixed label mapping when a new one label appears

Closes #92

chris-m-h commented on api/audit/events_api.go:
```diff
@@ -28,7 +28,7 @@ func NewEventService(concertoService utils.ConcertoService) (*EventService, erro
 	log.Debug("GetEventList")
 
-	data, status, err := cl.concertoService.Get("/v1/audit/events")
+	data, status, err := cl.concertoService.Get("/v2/audit/events")
```
Just a side-note: Could the version not be a part of the service configuration?

pcantera replied:
Sure, i was working on it for a time. Finally i decided not to include in current issue. (but my idea is to unify it in close future to avoid this)
I have an eye on it.

sjpatino commented on api/labels/labels_api_mocked.go:
```diff
@@ -383,7 +383,7 @@ func RemoveLabelMocked(t *testing.T, labelIn *types.Label) {
 	// call service
 	resourceID := "5b5074735f7c880ad9c6bbce"
-	cs.On("Delete", fmt.Sprintf("v1/labels/%s/resources/%s/%s", labelIn.ID, labelIn.ResourceType, resourceID)).Return(dIn, 204, nil)
+	cs.On("Delete", fmt.Sprintf("/v2/labels/%s/resources/%s/%s", labelIn.ID, labelIn.ResourceType, resourceID)).Return(dIn, 204, nil)
```
Why have you changed v1 for /v2?

pcantera replied:
Well, it should be "v2" and the slash is included due to normalization, canonical way should be.
The approach now,  trying to normalize the url/paths. every path should have an initial slash.

And for base domain, it could be provided by definition or not; but in any case, i applied piece of code to remove if exist at the end of base URL.
So with this way, the URLs are well formed and not repeating // in the path

sjpatino reviewed (approved):
It seems right ;)

chris-m-h commented on utils/format/textformatter.go:
```diff
@@ -33,6 +33,8 @@ func (f *TextFormatter) printItemAux(w *tabwriter.Writer, item interface{}) erro
 		showTags := strings.Split(it.Type().Field(i).Tag.Get("show"), ",")
 		if !utils.Contains(showTags, "noshow") {
 			switch it.Field(i).Type().String() {
+			case "time.Time":
```
Not sure I understand what we are doing here

pcantera replied:
Basically in previous PR, the print formatter required changes to support nested structures (as "Labelables" ones), then the code was re-factored and updated  in order to support it (The old code made integration and compatibility difficult).

But after that while testing a  command battery , this "time" type case was not included or not managed as it should be (so fixed here).

Probably in the future, formatter will be improve a bit more.

chris-m-h reviewed (approved):
Looks good to me. Just 2 comments

pcantera committed:
Make friendly usage for run_list recipes (issue #92)

Closes #92

pcantera committed:
Remove Variables property from template (issue #92)

Closes #92

pbanos reviewed (approved):
Approved!
